tests/Doctrine/Tests/ORM/Functional/IdentityMapTest.php

```diff
@@ -254,5 +255,50 @@ public function testCollectionValuedAssociationIdentityMapBehaviorWithRefresh()
         // Now the collection should be refreshed with correct count
         $this->assertEquals(4, count($user2->getPhonenumbers()));
     }
+
+    /**
+     * @group HashCollision
+     */
+    public function testHashCollision() {
+        $user = new CmsUser();
+        $user->username = "Test";
+        $user->name     = "Test";
+        $this->_em->persist($user);
+        $this->_em->flush();
+
+        $articles = [];
+        for($i=0;$i<100;$i++) {
```

Comment:
Can we reproduce this with just two?

Reply:
The problem is the underlying GC implementation. With PHP 5.6, it is triggering the assert at the 85th entity. With PHP7, it's at the 100th. I think the problem is based on what the internal implementation of PHP is choosing among the available memory slots. When I try with 2, nothing happen.

Reply:
PHP 5.6 is generally not relevant - what's important is to verify PHP 7. The fact that it acts on the very last iteration, yet we can't reduce the iteration count to just 2 means that we don't fully understand the problem yet.

Reply:
Well, we're talking about how PHP is recycling memory. What if PHP is storing available memory slots into an hashmap? It would result into something more or less random, this is why having a lot of slots having the same size being available (due to recycling) is making this error to happens more frequently. This unit tests is just showing the problem, but there's too much random factor in this bug. Keep in mind that this is happening, and why: the documentation about spl_object_hash is clear: ids are recycled, thus not unique in time. My fix is pretty simple, it just ensure the UOW is still holding a reference on managed entities to avoid this pitfall. To be honest, I don't understand why we need a unit tests for this case, this is part of the way PHP is handling memory. And we DO have a unit test that is showing the issue, what a chance :). I think we do understand what is happening here, this is why the unit test is "statically working". I think this unit test should NOT be merged into the core as PHP might change the way it handles memory in the future. This fix is important as long as the UOW is using spl_object_hash (and spl_object_id for PHP7).